druntime/src/object.d

```diff
@@ -3781,6 +3781,22 @@ template RTInfo(T)
         enum RTInfo = rtinfoNoPointers;
     else
         enum RTInfo = RTInfoImpl!(pointerBitmap).ptr;
+
+    static if (is (T == struct) && is(typeof(T.init) == T))
+    {
+        /// Verify that `destroy` works with `T.init`.
+        /// TODO: explain why
```

Comment:
cc @pbackus could you please provide a detailed explanation here?

Reply:
Short answer: I was wrong. Close this.

Long answer: the requirement I was thinking of is that destroying `T.init` must not cause *undefined behavior*, because `T.init` is always available in `@safe` code. But that does not mean it has to succeed—it could still throw an exception, enter an infinite loop, terminate the process, etc.

Reply:
> Long answer: the requirement I was thinking of is that destroying `T.init` must not cause _undefined behavior_, because `T.init` is always available in `@safe` code. But that does not mean it has to succeed—it could still throw an exception, enter an infinite loop, terminate the process, etc.

I don't think you can even rely on `T.init` always being available in `@safe` code due to it still being possible to define an `init()` method, which may be `@system`.

Reply:
Even if a type defines an `init()` method, you can still obtain its `.init` value by using something like `(T[1]).init`.

There is currently no reliable way to *prevent* `@safe` code from accessing a type's `.init` value.

Reply:
> There is currently no reliable way to _prevent_ `@safe` code from accessing a type's `.init` value.

A `@system` init field is also legal - `struct Foo { @system noreturn init; }`, for example. The `(T[1]).init` trick doesn't work around that one. It's only a deprecation right now, so TECHNICALLY it's allowed... for now.

Reply:
This works for me, even with `-preview=systemVariables` enabled:

```d
struct S
{
    int n = 123;
    @disable this();
    @system noreturn init;
}

void main() @safe
{
    auto a = (S[1]).init;
    assert(a[0].n == 123); // passes
}
```

Reply:
Hmm:
```d
struct S
{
    int n = 123;
    //@disable this();
    @system static immutable S init;
}

void main() @safe
{
    auto a = S.init;
    assert(a.n == 123); // passes
}
```
successfully compiles with `-preview=systemVariables` which seems like a bug to me

Reply:
It's definitely a bug, but it's also irrelevant to this PR which—again—should be closed.